ceno_zkvm/src/scheme/cpu/mod.rs

```diff
@@ -190,6 +215,109 @@ impl CpuTowerProver {
+                        let x3 = &SymbolicSepticExtension::new(
+                            x3.into_iter()
+                                .map(|x| expr_builder.lift(Either::Left(x)))
+                                .collect(),
+                        );
+                        let y3 = &SymbolicSepticExtension::new(
+                            y3.into_iter()
+                                .map(|y| expr_builder.lift(Either::Left(y)))
+                                .collect(),
+                        );
+                        let x1_prime_expr = SymbolicSepticExtension::new(
+                            x1_prime
+                                .iter_mut()
+                                .map(|x| expr_builder.lift(x.to_either()))
+                                .collect(),
+                        );
+                        let y1_prime_expr = SymbolicSepticExtension::new(
+                            y1_prime
+                                .iter_mut()
+                                .map(|y| expr_builder.lift(y.to_either()))
+                                .collect(),
+                        );
+                        let x2_prime_expr = SymbolicSepticExtension::new(
+                            x2_prime
+                                .iter_mut()
+                                .map(|x| expr_builder.lift(x.to_either()))
+                                .collect(),
+                        );
+                        let y2_prime_expr = SymbolicSepticExtension::new(
+                            y2_prime
+                                .iter_mut()
+                                .map(|y| expr_builder.lift(y.to_either()))
+                                .collect(),
+                        );
+
+                        // layer i: x3', y3', x1', y1', x2', y2', each has `i` variables
+                        // we copy the first half of x3 to x1', 2nd half to x2' and
+                        // copy the first half of y3 to y1', 2nd half to y2'.
+                        //
+                        // x1'[b] = x3[0,b], y1'[b] = y3[0,b]
+                        // x2'[b] = x3[1,b], y2'[b] = y3[1,b]
+                        //
+                        // layer i+1: x3, y3, x1, y1, x2, y2, each has `i+1` variables
+                        // we requires the elliptic curve addition constraints hold at layer i+1.
+                        // 1. 0 = \sum_b eq(rt,b) * ((x3 + x1 + x2) * (x2 - x1)^2 - (y2 - y1)^2)
+                        exprs.extend(
+                            (((x3 + x1 + x2) * (x2 - x1) * (x2 - x1) - (y2 - y1) * (y2 - y1))
+                                * &eq_expr)
+                                .to_exprs(),
+                        );
+                        // 2. 0 = \sum_b eq(rt,b) * ((y3 + y1) * (x2 - x1) - (y2 - y1) * (x1 - x3))
+                        exprs.extend(
+                            (((y3 + y1) * (x2 - x1) - (y2 - y1) * (x1 - x3)) * &eq_expr).to_exprs(),
+                        );
```

Comment:
Since we cannot do inversion inside a sumcheck, to verify $(x_3, y_3) = (x_1, y_1) + (x_2, y_2)$ with the formulas
1. $x_3 = (y_2 - y_1)^2 / (x_2 - x_1)^2 - x_1 - x_2$;
2. $y_3 = (y_2 - y_1) / (x_2 - x_1) * (x_1 - x_3) - y_1$.

we convert the inversion by product, then we need to copy the result $(x_3, y_3)$ to the output layer.

Therefore, we require that $(x'_1, y'_1)$ at layer $i$ must be the 1st half of $(x_3, y_3)$, and $(x'_2, y'_2)$ at layer $i$ must be the 2nd half of $(x_3, y_3)$.